### PR TITLE
fix(*): status codes returning from backend after registration

### PR DESCRIPTION
**Pull request by Kif-Fando**
Backend returning status code 500 after successful registration

### CODE REVIEW

File: public/src/pages/auth/Register/Register.tsx [hunk 1]
```diff
@@ -9,30 +9,64 @@ const defaultUser: RegistrationUser = {
...
 };
 
 export const Register: FC = () => {
   const [form, setForm] = useState<RegistrationUser>(defaultUser);
   const { email, firstName, lastName, password } = form;
 
   const [regResponse, setRegResponse] = useState<RegistrationUser | null>();
+  const [errorText, setErrorText] = useState<string | null>();
+
+  const [mode, setMode] = useState<LoginFormMode>(LoginFormMode.BASIC);
```
Comment (tamirGer) [inline public/src/pages/auth/Register/Register.tsx:23]: Change to `authMode`, `setAuthMode`, for more readable code

File: public/src/pages/auth/Register/Register.tsx [hunk 2]
```diff
@@ -9,30 +9,64 @@ const defaultUser: RegistrationUser = {
...
 
   const onInput = ({ target }: { target: EventTarget | null }) => {
     const { name, value } = target as HTMLInputElement;
     setForm({ ...form, [name]: value });
   };
 
+  const onSelectMode = ({ target }: { target: EventTarget | null }) => {
```
Comment (tamirGer) [inline public/src/pages/auth/Register/Register.tsx:30]: Change to `onAuthModeChange`

File: public/src/pages/auth/Register/Register.tsx [hunk 3]
```diff
@@ -9,30 +9,64 @@ const defaultUser: RegistrationUser = {
...
+    const { value } = target as HTMLSelectElement & { value: LoginFormMode };
+    setForm({ ...form, op: value });
+    setMode(value);
+    switch (value as LoginFormMode) {
```
Comment (tamirGer) [inline public/src/pages/auth/Register/Register.tsx:34]: Remove unnecessary switch statement

File: src/keycloak/keycloak.service.ts [hunk 4]
```diff
@@ -137,17 +141,40 @@ export class KeyCloakService implements OnModuleInit {
...
+      `${this.server_uri}/admin/realms/${this.realm}/users?email=${email}`,
+      {
+        headers: {
+          'Content-Type': 'application/json',
+          Authorization: `${token_type} ${access_token}`,
+        },
+        responseType: 'json',
+      }
+    );
+    if (existingUser[0]) {
```
Comment (tamirGer) [inline src/keycloak/keycloak.service.ts:161]: Change to
`return existingUser[0] ? true : false`

File: src/keycloak/keycloak.service.ts [hunk 5]
```diff
@@ -137,17 +141,40 @@ export class KeyCloakService implements OnModuleInit {
...
+    return false;
+  }
+
   public async registerUser({
     firstName,
     lastName,
     email,
     password,
-  }: RegisterUserData): Promise<void> {
+  }: RegisterUserData): Promise<any> {
```
Comment (tamirGer) [inline src/keycloak/keycloak.service.ts:172]: Please add a proper return type, avoid using type `any`

File: src/users/users.controller.ts [hunk 6]
```diff
@@ -163,8 +164,14 @@ export class UsersController {
   })
   async createUser(@Body() user: CreateUserRequest): Promise<UserDto> {
     try {
-      this.logger.debug(`Create a user: ${user}`);
-
+      this.logger.debug(`Create a basic user: ${user}`);
+      const userIsExist = await this.usersService.findByEmail(user.email);
```
Comment (tamirGer) [inline src/users/users.controller.ts:168]: Change to `userExists`

File: src/users/users.controller.ts [hunk 7]
```diff
@@ -173,20 +180,44 @@ export class UsersController {
...
+  })
+  @ApiResponse({
+    type: UserDto,
+    status: 200,
+  })
+  async createOIDCUser(@Body() user: CreateUserRequest): Promise<any> {
+    try {
+      this.logger.debug(`Create a OIDC user: ${user}`);
+
+      const userIsExist = await this.keyCloakService.isUserExists({
```
Comment (tamirGer) [inline src/users/users.controller.ts:201]: Change to `userExists`

File: public/src/api/makeApiRequest.ts [hunk 8]
```diff
@@ -24,6 +24,12 @@ export function makeApiRequest<T>(
             'Authentication failed, please check your credentials and try again'
         };
       }
+      if (error.response.status === 409) {
```
Comment (tamirGer) [inline public/src/api/makeApiRequest.ts:27]: Please change this if statement and the previous one into one switch statement

**Review by tamirGer — CHANGES_REQUESTED** [review]

File: src/keycloak/keycloak.service.ts [hunk 9]
```diff
@@ -137,17 +141,37 @@ export class KeyCloakService implements OnModuleInit {
...
+    return existingUser[0] ? true : false;
+  }
+
   public async registerUser({
     firstName,
     lastName,
     email,
     password,
-  }: RegisterUserData): Promise<void> {
+  }: RegisterUserData): Promise<any> {
```
Comment (tamirGer) [inline src/keycloak/keycloak.service.ts:169]: Please change type `any` to a specific type

**Review by tamirGer — CHANGES_REQUESTED** [review]

**Comment by derevnjuk** [comment]
@Kif-Fando Please adjust PR title and body according to contributing guide. Also, don't put references to JIra if you work on open-source projects.

File: src/keycloak/keycloak.service.ts [hunk 10]
```diff
@@ -137,17 +142,37 @@ export class KeyCloakService implements OnModuleInit {
...
+      `${this.server_uri}/admin/realms/${this.realm}/users?email=${email}`,
+      {
+        headers: {
+          'Content-Type': 'application/json',
+          Authorization: `${token_type} ${access_token}`,
+        },
+        responseType: 'json',
+      }
+    );
+    return existingUser[0] ? true : false;
```
Comment (derevnjuk) [inline src/keycloak/keycloak.service.ts:162]:
```suggestion
    const [existingUser] = await this.httpClient.get(
      `${this.server_uri}/admin/realms/${this.realm}/users?email=${email}`,
      {
        headers: {
          'Content-Type': 'application/json',
          Authorization: `${token_type} ${access_token}`,
        },
        responseType: 'json',
      }
    );
    return !!existingUser;
```

File: public/src/api/httpClient.ts [hunk 11]
```diff
@@ -51,7 +51,7 @@ export function postSubscriptions(email: string): Promise<any> {
 
 export function postUser(data: RegistrationUser): Promise<any> {
   return makeApiRequest({
-    url: ApiUrl.Users,
+    url: `${ApiUrl.Users}/${data.op}`,
```
Comment (derevnjuk) [inline public/src/api/httpClient.ts:54]: What is the purpose of these changes?

File: src/users/users.controller.ts [hunk 12]
```diff
@@ -173,20 +180,44 @@ export class UsersController {
           user.password,
         ),
       );
+      return newUser;
+    } catch ({ response = 'Something went wrong', status = 500 }) {
+      throw new HttpException(response, status)
+    }
```
Comment (derevnjuk) [inline src/users/users.controller.ts:186]: Use the `message` field instead of `response`. Avoid using destructuring assignment in `catch` statement since `err` has unknown or `any` type depending on tsc settings
 
```suggestion
    } catch (err) {
      throw new HttpException(err.message ?? 'Something went wrong', err.status ?? 500)
    }
```

**Review by derevnjuk — CHANGES_REQUESTED** [review]

**Review by derevnjuk — APPROVED** [review]

**Review by tamirGer — APPROVED** [review]
